source/Server/IntegratedAuthentication/IntegratedChallengeCoordinator.cs

```diff
@@ -64,10 +66,29 @@ public IntegratedChallengeTrackerStatus SetupResponseIfChallengeHasNotSucceededY
                         Secure = state?.UsingSecureConnection ?? false
                     });
                 
-                // we pass back to the original link here (e.g. the deep link that originally triggered the sign in)
+                // we pass back to the original link here if it is a trusted url (e.g. the deep link that originally triggered the sign in)
                 // we normally wouldn't do this without the user being authenticated, but given we know they aren't
                 // authenticated we'll redirect back and rely on the sign in page kicking in again and seeing the cookie.
-                context.Response.Redirect(stateRedirectAfterLoginTo);
+                // If the caller has provided a redirect after successful login, we need to check it is a local address - preventing Open Redirection attacks
+                if (!string.IsNullOrWhiteSpace(state?.RedirectAfterLoginTo))
```

Comment:
[Minor] Looks like we already have a local variable `stateRedirectAfterLoginTo` for this and a few subsequent references to the same property.